robosuite/controllers/ee_imp.py

```diff
@@ -0,0 +1,180 @@
+                                                 set_ori=set_ori)
+        self.goal_pos = set_goal_position(scaled_delta[:3],
+                                          self.ee_pos,
+                                          position_limit=self.position_limits,
+                                          set_pos=set_pos)
+
+        if self.interpolator_pos is not None:
+            self.interpolator_pos.set_goal(self.goal_pos)
+            print("Got here!")
+
+        if self.interpolator_ori is not None:
+            self.ori_ref = np.array(self.ee_ori_mat)  # reference is the current orientation at start
+            self.interpolator_ori.set_goal(orientation_error(self.goal_ori, self.ori_ref))  # goal is the total orientation error
+            self.relative_ori = np.zeros(3)  # relative orientation always starts at 0
+
+    def run_controller(self, action=None):
+        # First, update goal if action is not set to none
+        # Action will be interpreted as delta value from current
+        if action is not None:
+            self.set_goal(action)
+        else:
+            self.update()
+
+        desired_pos = None
+        # Only linear interpolator is currently supported
+        if self.interpolator_pos is not None:
+            # Linear case
+            if self.interpolator_pos.order == 1:
+                desired_pos = self.interpolator_pos.get_interpolated_goal(self.ee_pos)
+            else:
+                # Nonlinear case not currently supported
+                pass
+        else:
+            desired_pos = np.array(self.goal_pos)
+
+        if self.interpolator_ori is not None:
+            # relative orientation based on difference between current ori and ref
+            self.relative_ori = orientation_error(self.ee_ori_mat, self.ori_ref)
+
+            interpolated_results = self.interpolator_ori.get_interpolated_goal(self.relative_ori)
+            ori_error = interpolated_results[0:3]
+        else:
+            desired_ori = np.array(self.goal_ori)
+            ori_error = orientation_error(desired_ori, self.ee_ori_mat)
+
+        position_error = desired_pos - self.ee_pos
+        vel_pos_error = -self.ee_pos_vel
+        desired_force = (np.multiply(np.array(position_error), np.array(self.kp[0:3]))
+                         + np.multiply(vel_pos_error, self.kv[0:3]))
+
+        vel_ori_error = -self.ee_ori_vel
+        desired_torque = (np.multiply(np.array(ori_error), np.array(self.kp[3:6]))
+                          + np.multiply(vel_ori_error, self.kv[3:6]))
+
+        lambda_full, lambda_pos, lambda_ori, nullspace_matrix = opspace_matrices(self.mass_matrix,
+                                                                                 self.J_full,
+                                                                                 self.J_pos,
+                                                                                 self.J_ori)
+        if self.uncoupling:
```

Comment:
could use a comment on what uncoupling does. In general, for this controller implementation, it would be great to reference or add equations in comments for what's going on in the OSC computation (for example, tau = J.T * F, F = -kp(x - x_des) -kv(xdot), etc.) - otherwise people will be lost when reading this code and trying to understand what happens

Reply:
done